File: app-java/src/main/java/com/kotlin/app_java/MainActivity.java
package com.kotlin.app_java;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity {

    private WeatherAdapter adapter;
    private TextView counter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        counter = findViewById(R.id.textView);

        initAdapter();

        setTitle();

        addFilter();

    }

    void initAdapter() {
        RecyclerView recyclerView = findViewById(R.id.recyclerView);

        adapter = new WeatherAdapter();

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(this, DividerItemDecoration.HORIZONTAL);

        dividerItemDecoration.setDrawable(getDrawable(R.drawable.divider));

        recyclerView.addItemDecoration(dividerItemDecoration);

        List<Weather> weatherList = Repository.getInstance().getWeatherList();
        adapter.setWeatherList(weatherList);
        setCounter(weatherList.size());
    }

    void setTitle() {
        getSupportActionBar().setTitle(Utils.getFullName(
                this.getPackageName(),
                this.getString(R.string.app_name)
        ));
    }

    void addFilter() {
        ((EditText) findViewById(R.id.editText)).addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                applyFilter(s.toString());
            }
        });
    }

    void applyFilter(String s) {
        List<Weather> filteredWeatherList = new ArrayList<>();

        for (Weather weather : Repository.getInstance().getWeatherList()) {
            if (weather.getTown().toLowerCase().contains(s.toLowerCase())) {
                filteredWeatherList.add(weather);
            }
        }
        adapter.setWeatherList(filteredWeatherList);
        setCounter(filteredWeatherList.size());
    }

    void setCounter(Integer count) {
        if (count == 0) {
            counter.setText("Нет городов");
        } else {
            counter.setText(String.format("Городов: %d", count));
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/com/kotlin/gbapplication2308/MainActivity.kt
package com.kotlin.gbapplication2308

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var adapter: WeatherAdapter
    private lateinit var counter: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        supportActionBar?.title = getFullName(packageName, getString(R.string.app_name))

        counter = findViewById(R.id.textView)


        counter.setOnClickListener {
            Toast.makeText(this@MainActivity, "sdfsdf", Toast.LENGTH_LONG).show()
        }

        initAdapter()

        addFilter()
    }

    private fun initAdapter() {
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)

        adapter = WeatherAdapter()

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter
        adapter.weatherList = Repository.weatherList
        setCounter(adapter.weatherList.size)
    }

    fun addFilter() {
        findViewById<EditText>(R.id.editText).addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                applyFilter(s.toString())
            }
        })
    }

    fun applyFilter(s: String) {

        val filteredWeatherList: MutableList<Weather> = mutableListOf()

        for (weather in Repository.weatherList) {
            if (weather.town.contains(other = s, ignoreCase = true)) {
                filteredWeatherList.add(weather)
            }
        }

        adapter.weatherList = filteredWeatherList
        setCounter(filteredWeatherList.size)
    }

    private fun setCounter(count: Int) {
        counter.text = if (count == 0) "Нет городов" else "Городов: $count"
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            true
        } else {
            super.onOptionsItemSelected(item)
        }

}